### ISSUE TITLE
[BUG] "Incoming message has no 'type' attribute"

### ISSUE THREAD

**Issue by mauroborghesi** [hishnash/djangochannelsrestframework#23]
**Multiple connections not working on Observer**
We are working on a Django project, we are trying to implement channels connections but when we try to observe our Model instance change on multiple connections we end up viewing the change only on one of them, reporting on the other "DISCONNECTED" and `ValueError("Incoming message has no 'type' attribute")`.
This is our Consumer:
```python
class BoardObserverConsumer(ObserverModelInstanceMixin, GenericAsyncAPIConsumer):
    queryset = Board.objects.all()
    serializer_class = BoardChannelSerializer
```
and this is the Serializer:
```python
class BoardChannelSerializer(HyperChannelsApiModelSerializer):
    class Meta:
        model = Board
        fields = ('@id', 'name', 'players_count')
```
The version we are currently using is 0.0.3 as it is required by HyperMediaChannels.

**To Reproduce**
We connected various browsers with "in incognito" mode to the same localhost.

**Expected behavior**
We expect to be able to see the changes from all connections indistinguishably.

**LOG**
```
Exception inside application: Incoming message has no 'type' attribute
Traceback (most recent call last):
  File "/Users/myusername/Documents/Development/AmbVirt/lib/python3.7/site-packages/channels/sessions.py", line 183, in __call__
    return await self.inner(receive, self.send)
  File "/Users/myusername/Documents/Development/AmbVirt/lib/python3.7/site-packages/channels/middleware.py", line 41, in coroutine_call
    await inner_instance(receive, send)
  File "/Users/myusername/Documents/Development/AmbVirt/lib/python3.7/site-packages/channelsmultiplexer/demultiplexer.py", line 54, in __call__
    await future
  File "/Users/myusername/Documents/Development/AmbVirt/lib/python3.7/site-packages/channels/consumer.py", line 59, in __call__
    [receive, self.channel_receive], self.dispatch
  File "/Users/myusername/Documents/Development/AmbVirt/lib/python3.7/site-packages/channels/utils.py", line 51, in await_many_dispatch
    await dispatch(result)
  File "/Users/myusername/Documents/Development/AmbVirt/lib/python3.7/site-packages/channels/consumer.py", line 71, in dispatch
    handler = getattr(self, get_handler_name(message), None)
  File "/Users/myusername/Documents/Development/AmbVirt/lib/python3.7/site-packages/channels/consumer.py", line 19, in get_handler_name
    raise ValueError("Incoming message has no 'type' attribute")
ValueError: Incoming message has no 'type' attribute
WebSocket DISCONNECT /ws/channel/ [127.0.0.1:59317]
WebSocket HANDSHAKING /ws/channel/ [127.0.0.1:59354]
WebSocket CONNECT /ws/channel/ [127.0.0.1:59354]
```

 - OS: macOS 10.15.4 but same issue reproduced on Windows 10 Home

**Comment by hishnash** [comment]
@mauroborghesi  thank you for reporting this. This does remind me that i need to update `HyperMediaChannels` for this latest version.

However this issue seems unlikely to be related to that, 

it would normally be hitting this code that sets the `type` field. 

https://github.com/hishnash/djangochannelsrestframework/blob/v0.0.3/djangochannelsrestframework/observer/observer.py#L253-L261

Would you be able to set a breakpoint here to assert that it is calling this method?

What channel layer are you using?

**Comment by mauroborghesi** [comment]
Hi,

Regarding the channel layer, we are using Redis.

We tried to breakpoint the code and found that `message['type']` is created and is `handle.instance.change`. Though, this only occurs once, which is for the first connection; should the code step through the breakpoint once for each connection to the channel? Or should it create more of them - one for each connection?

Thank you very much.

**Comment by hishnash** [comment]
the code should step through there once for every update/addition made to the model. It should not be affected by connections opening/closing.

**Comment by mauroborghesi** [comment]
Then it seems to work fine. But why does the first browser read the `message['type']` and not any other? Do you have any guess/suggestion?

**Comment by hishnash** [comment]
are you doing any other message sending to channels groups in your code?

**Comment by mauroborghesi** [comment]
Other messages with action `retrieve` or `update` are working fine

**Comment by hishnash** [comment]
strange, I will try to get an update to `channelsmultiplexer` so that it can use the latest channels. Im not 100% sure if the `message` that does no have the `type` fields Is coming over redis or coming up the middleware stack from Daphne.

**Comment by MarcoGiuliani89** [comment]
Hi I'm working with Mauro.
I see where is the problem.
We have multiple clients and an observed instance.

I debug channels.consumer in `get_handler_name` and your `djangochannelsrestframework.observer.generics` in 
```python 
@_GenericModelObserver    async def handle_instance_change(self, message, **kwargs):
```

When I update the observed instance I first have a `get_handler_name` with message: 
`{'pk': 2, 'type': 'handle.instance.change', 'action': 'update'}`, 
then your `handle_instance_change` where the message change from `{'pk': 2, 'type': 'handle.instance.change', 'action': 'update'}` to `{'pk': 2}`
and finally we have others `get_handler_name` (one fore each other clients after the first one) with new message `{'pk': 2}`, so the error "Incoming message has no 'type' attribute"

In this way first client recive the update and other disconnect with the error.

Why you need to pop 'type' and 'action' in `handle_instance_change`?
Are we using your package in wrong way?

**Comment by hishnash** [comment]
```python
@_GenericModelObserver   
async def handle_instance_change(self, message, **kwargs):
```

should be called on each instance that has subscribed, it will then call `handle_observed_action` that will (if you have permission) send a message down the websocket.

but I think I understand the issue here. the dict is mutable ... ! so the pop id modifying the dict :) thank you for finding this I will get an update ASAP.

**Comment by hishnash** [comment]
I have released v0.0.6 i am just updating `hypermediachannels` so that it will let you use this new version. Thankyou for finding this issue, in production i have mostly been using a differnt channel layer that handed out fresh instances of the message dict to each consumer.

**Comment by hishnash** [comment]
hypermediachannels v0.0.5 is out.
